owiddata/01.OWID.basicStats.py

```diff
@@ -0,0 +1,89 @@
+import pandas as pd
+import os
+import datetime
+import argparse
+from jsonFunctions import *
+
+def convert_date(git_date):
+    '''Reformat git commit style datetimes (ISO 8601) to Month DD, YYYY.
+    Throws a ValueError if git_date cannot be parsed as an ISO 8601 datetime.
+    '''
+    # 'Z' indicates Coordinated Universal Time (UTC)
+    # Replace with '+00:00', which is the UTC representation recognized
+    # by the parser
+    # https://en.wikipedia.org/wiki/ISO_8601#Coordinated_Universal_Time_(UTC)
+    git_date = git_date.replace('Z', '+00:00')
+
+    # Remove the leading zero of the day
+    # Assumes the year will not begin with 0
+    return datetime.datetime.fromisoformat(git_date).strftime('%B %d, %Y').replace(' 0', ' ')
+
+def main(args):
+    # Create dictionary that will be exported as JSON
+    owid_stats = dict()
+
+    # Download data from a specific commit if the environment variable is set,
+    # otherwise default to master
+    commit = 'master'
+    if('OWID_COMMIT_SHA' in os.environ):
+        commit = os.environ['OWID_COMMIT_SHA']
+        owid_stats['owid_commit_sha'] = os.environ['OWID_COMMIT_SHA']
+    if('OWID_COMMIT_DATE' in os.environ):
+        owid_stats['owid_commit_date'] = os.environ['OWID_COMMIT_DATE']
+        owid_stats['owid_commit_date_pretty'] = convert_date(os.environ['OWID_COMMIT_DATE'])
+
+    # Retrieve data from OWID
+    locations_url = f'https://raw.githubusercontent.com/owid/covid-19-data/{commit}/public/data/vaccinations/locations.csv'
+    vaccine_locations = pd.read_csv(locations_url, error_bad_lines='skip')
+
+    numbers_url = f'https://raw.githubusercontent.com/owid/covid-19-data/{commit}/public/data/vaccinations/vaccinations.csv'
+    vaccine_nums = pd.read_csv(numbers_url, error_bad_lines='skip')
+
+    #manufacturer_url = f'https://raw.githubusercontent.com/owid/covid-19-data/{commit}/public/data/vaccinations/vaccinations-by-manufacturer.csv'
+    #vaccine_manf = pd.read_csv(manufacturer_url, error_bad_lines='skip')
+
+    # Pull up-to-date statistics from data
+    vaccine_nums['date'] = pd.to_datetime(vaccine_nums['date'])
+
+    owid_stats["owid_most_recent_date"] = vaccine_nums['date'].max().strftime('%B %d, %Y').replace(' 0', ' ')
+
+    owid_stats["owid_total_vaccinations"] = \
+        str("{:,}".format(round(vaccine_nums[vaccine_nums["location"] == "World"].
+                                loc[vaccine_nums["date"] ==
+                                    owid_stats["owid_most_recent_date"],
+                                    "total_vaccinations"].item()/1000000000))) + \
+        " billion"
```

Comment:
Maybe look into [plydata](https://plydata.readthedocs.io/en/stable/). Could turn the above into:

```python
owid_stats = (
  owid_stats
  >> ply.inner_join(vaccine_nums, left_on="owid_most_recent_date", right_on="date")
  >> ply.define(owid_total_vaccinations="total_vaccinations/1000000000")
  >> ply.define(owid_total_vaccinations = lambda x: x.owid_total_vaccinations.apply(lambda y: f"{y:,}"))
)
```

Reply:
Depending on the way you need to manipulate your data this package might be annoying; however, I do recommend checking it out as it is basically R tidyverse (pythonized).